### PR TITLE
Tests were developed with singleton driver and page object model patterns

### PR DESCRIPTION
**Pull request by igorchupin**

### CODE REVIEW

File: src/main/java/SingleDriver.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        return driver;
+    }
+
+    public static void close() {
+        if (driver != null) {
+            driver.close();
+        }
+    }
+
+    public static void clearBrowserData() {
```
Comment (EugenBorisik) [inline src/main/java/SingleDriver.java]: driver.manage().deleteAllCookies();

File: src/main/java/SingleDriver.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeDriver;
+
+import java.util.concurrent.TimeUnit;
+
+public class SingleDriver {
+    private static WebDriver driver;
+
+    public static WebDriver getInstance() {
```
Comment (EugenBorisik) [inline src/main/java/SingleDriver.java]: getInstance() method must return singleton instance (SingleDriver type)

File: src/main/java/SingleDriver.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                    driver.manage().timeouts().implicitlyWait(5, TimeUnit.SECONDS);
+                }
+            }
+        }
+        return driver;
+    }
+
+    public static void close() {
+        if (driver != null) {
+            driver.close();
```
Comment (EugenBorisik) [inline src/main/java/SingleDriver.java:34]: driver.close() just closes window, but driver != null after that. So you need to close and then assign null to driver

File: src/main/java/SingleDriver.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import org.openqa.selenium.By;
+import org.openqa.selenium.Keys;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeDriver;
+
+import java.util.concurrent.TimeUnit;
+
+public class SingleDriver {
+    private static WebDriver driver;
```
Comment (EugenBorisik) [inline src/main/java/SingleDriver.java]: Here must be instance of singleton class type

File: src/main/java/LoginPage.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+public class LoginPage {
+    private final WebDriver driver;
+    private final String loginPageURL = "https://mail.yandex.com/";
+    private final By enterButton = By.xpath("//*[@id=\"index-page-container\"]/div/div[2]/div/div/div[4]/a[2]");
+    private final By emailField = By.id("passp-field-login");
+    private final By loginButton = By.id("passp:sign-in");
+    private final By passwordField = By.id("passp-field-passwd");
+
+    public LoginPage(WebDriver driver) {
```
Comment (EugenBorisik) [inline src/main/java/LoginPage.java]: You have singleton for WebDriver, there is no needs to pass driver as param

File: src/main/java/LoginPage.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+
+public class LoginPage {
+    private final WebDriver driver;
+    private final String loginPageURL = "https://mail.yandex.com/";
+    private final By enterButton = By.xpath("//*[@id=\"index-page-container\"]/div/div[2]/div/div/div[4]/a[2]");
```
Comment (EugenBorisik) [inline src/main/java/LoginPage.java]: I will pretend that I didn't see this locator

File: src/main/java/LoginPage.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    public LoginPage sendUsername(String username) {
+        driver.findElement(emailField).sendKeys(username);
+        return this;
+    }
+
+    public LoginPage sendPassword(String password) {
+        driver.findElement(passwordField).sendKeys(password);
+        return this;
+    }
```
Comment (EugenBorisik) [inline src/main/java/LoginPage.java]: I hope you don't need 1-line methods, but at least make them private

File: src/main/java/LoginPage.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+
+public class LoginPage {
+    private final WebDriver driver;
+    private final String loginPageURL = "https://mail.yandex.com/";
+    private final By enterButton = By.xpath("//*[@id=\"index-page-container\"]/div/div[2]/div/div/div[4]/a[2]");
+    private final By emailField = By.id("passp-field-login");
+    private final By loginButton = By.id("passp:sign-in");
+    private final By passwordField = By.id("passp-field-passwd");
```
Comment (EugenBorisik) [inline src/main/java/LoginPage.java]: I think all locators can be constants

File: src/main/java/HomePage.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+public class HomePage {
+    private final WebDriver driver;
+    private final String EXPECTED_NAME = "Inbox — Yandex.Mail";
+    private final By userAccountName = By.xpath("//div[@class='legouser__menu-header']//span[@class='user-account__name']");
+    private final By userIcon = By.className("user-pic__image");
+    private final By signOutButton = By.linkText("Log out");
```
Comment (EugenBorisik) [inline src/main/java/HomePage.java]: Make them constants

File: src/main/java/HomePage.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+    public HomePage(WebDriver driver) {
```
Comment (EugenBorisik) [inline src/main/java/HomePage.java]: No needs to pass WebDriver instance as param

File: src/main/java/HomePage.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+public class HomePage {
+    private final WebDriver driver;
+    private final String EXPECTED_NAME = "Inbox — Yandex.Mail";
```
Comment (EugenBorisik) [inline src/main/java/HomePage.java]: Expected values (test data) must be in test classes

File: src/main/java/HomePage.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    private final By signOutButton = By.linkText("Log out");
+
+    public HomePage(WebDriver driver) {
+        this.driver = driver;
+    }
+
+    public HomePage openUserMenu() {
+        driver.findElement(userIcon).click();
+        return this;
+    }
```
Comment (EugenBorisik) [inline src/main/java/HomePage.java]: I think you don't need this 1-line method, but at least make it private

File: src/main/java/HomePage.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+    public String getAccountName() {
+        openUserMenu();
+        WebDriverWait wait = new WebDriverWait(driver, 10);
+        wait.until(ExpectedConditions.visibilityOfElementLocated(userAccountName));
+        return driver.findElement(userAccountName).getText();
+    }
+
+    public ChooseAccountPage clickSignOut() {
```
Comment (EugenBorisik) [inline src/main/java/HomePage.java]: signOut is better name

File: src/main/java/ChooseAccountPage.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+public class ChooseAccountPage {
+
+    private final WebDriver driver;
+    private final By youAreLoggedOutText = By.xpath("//h3[@class='AuthAccountList-unathorizedAccountsTitle']");
+    private final By accountsList = By.xpath("//div/a[@class='AuthAccountListItem']");
+    private final By currentAccountName = By.xpath("//span[@class='CurrentAccount-displayName']");
```
Comment (EugenBorisik) [inline src/main/java/ChooseAccountPage.java]: Can be constants

File: src/main/java/ChooseAccountPage.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+    public ChooseAccountPage(WebDriver driver) {
```
Comment (EugenBorisik) [inline src/main/java/ChooseAccountPage.java]: No needs to pass WebDriver instance as parameter

File: src/main/java/ChooseAccountPage.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        this.driver = driver;
+    }
+
+    public String getLogoutText() {
```
Comment (EugenBorisik) [inline src/main/java/ChooseAccountPage.java]: Don't like this name

File: src/main/java/ChooseAccountPage.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+        WebElement account = driver.findElement(currentAccountName);
+        account.click();
+        WebDriverWait wait = new WebDriverWait(driver, 10);
+        wait.until(ExpectedConditions.visibilityOfElementLocated(youAreLoggedOutText));
+        return driver.findElement(youAreLoggedOutText).getText();
+    }
+
+    public List<WebElement> getAccountsList() {
```
Comment (EugenBorisik) [inline src/main/java/ChooseAccountPage.java]: If public method in page object class returns WebElement - this signals that here you have error. Sense of page object - encapsulate elements and business functions. If you return WebElement, then you provide direct access to element on the page

File: src/test/java/LoginLogoutTest.java [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        SingleDriver.close();
+    }
+
+    @Test
+    @Tag("yandex_login_test")
+    @DisplayName("Login in mail test")
+    void yandexLoginTest() {
+        loginPage.loginToMail(username, password);
+        WebDriverWait wait = new WebDriverWait(driver, 5);
+        wait.until(ExpectedConditions.titleContains(homePage.getPageName()));
```
Comment (EugenBorisik) [inline src/test/java/LoginLogoutTest.java]: This waiter can be in HomePage constructor. Waiters must be in page objects.

File: src/test/java/LoginLogoutTest.java [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class LoginLogoutTest {
+    private static WebDriver driver;
+    private LoginPage loginPage = new LoginPage(driver);
+    private HomePage homePage = new HomePage(driver);
+    private ChooseAccountPage chooseAccountPage = new ChooseAccountPage(driver);
```
Comment (EugenBorisik) [inline src/test/java/LoginLogoutTest.java]: Remove this 2 lines

File: src/test/java/LoginLogoutTest.java [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    @AfterAll
+    public static void closeDriver() {
+        SingleDriver.close();
+    }
+
+    @Test
+    @Tag("yandex_login_test")
+    @DisplayName("Login in mail test")
+    void yandexLoginTest() {
+        loginPage.loginToMail(username, password);
```
Comment (EugenBorisik) [inline src/test/java/LoginLogoutTest.java]: HomePage homePage = loginPage.loginToMail(username, password);

File: src/test/java/LoginLogoutTest.java [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+                () -> assertEquals(username + "\n" + username + "@yandex.com", homePage.getAccountName(),
+                        "Incorrect user account name is shown")
+        );
+    }
+
+    @Test
+    @Tag("yandex_logout_test")
+    @DisplayName("Logout from mail test")
+    void yandexLogoutTest() {
+        loginPage.loginToMail(username, password);
```
Comment (EugenBorisik) [inline src/test/java/LoginLogoutTest.java]: HomePage homePage = loginPage.loginToMail(username, password);

File: src/test/java/LoginLogoutTest.java [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+        homePage.clickSignOut();
```
Comment (EugenBorisik) [inline src/test/java/LoginLogoutTest.java]: ChooseAccountPage chooseAccountPage = homePage.clickSignOut();

**Review by EugenBorisik — CHANGES_REQUESTED** [review]
Correct the same places in PR with Page Factory

File: src/main/java/SingleDriver.java (hunk 1 above)
Comment (igorchupin) [reply src/main/java/SingleDriver.java]: I tried it but it does not help

File: src/main/java/SingleDriver.java (hunk 2 above)
Comment (igorchupin) [reply src/main/java/SingleDriver.java]: Method was changed

File: src/main/java/SingleDriver.java (hunk 3 above)
Comment (igorchupin) [reply src/main/java/SingleDriver.java:34]: This was changed, thank you!

File: src/main/java/SingleDriver.java (hunk 4 above)
Comment (igorchupin) [reply src/main/java/SingleDriver.java]: This was added

File: src/main/java/LoginPage.java (hunk 5 above)
Comment (igorchupin) [reply src/main/java/LoginPage.java]: This was changed

File: src/main/java/LoginPage.java (hunk 6 above)
Comment (igorchupin) [reply src/main/java/LoginPage.java]: Locator was changed =)))

File: src/main/java/LoginPage.java (hunk 7 above)
Comment (igorchupin) [reply src/main/java/LoginPage.java]: One line methods were deleted

File: src/main/java/LoginPage.java (hunk 8 above)
Comment (igorchupin) [reply src/main/java/LoginPage.java]: All locators were renamed

File: src/main/java/HomePage.java (hunk 9 above)
Comment (igorchupin) [reply src/main/java/HomePage.java]: All locators were renamed

File: src/main/java/HomePage.java (hunk 10 above)
Comment (igorchupin) [reply src/main/java/HomePage.java]: This was changed

File: src/main/java/HomePage.java (hunk 11 above)
Comment (igorchupin) [reply src/main/java/HomePage.java]: This was changed

File: src/main/java/HomePage.java (hunk 12 above)
Comment (igorchupin) [reply src/main/java/HomePage.java]: Method was deleted

File: src/main/java/HomePage.java (hunk 13 above)
Comment (igorchupin) [reply src/main/java/HomePage.java]: Name was changed

File: src/main/java/ChooseAccountPage.java (hunk 14 above)
Comment (igorchupin) [reply src/main/java/ChooseAccountPage.java]: This was changed

File: src/main/java/ChooseAccountPage.java (hunk 15 above)
Comment (igorchupin) [reply src/main/java/ChooseAccountPage.java]: Constructor was changed

File: src/main/java/ChooseAccountPage.java (hunk 16 above)
Comment (igorchupin) [reply src/main/java/ChooseAccountPage.java]: Name was changed

File: src/main/java/ChooseAccountPage.java (hunk 17 above)
Comment (igorchupin) [reply src/main/java/ChooseAccountPage.java]: Actually, this method is not needed so I decided to remove it

File: src/test/java/LoginLogoutTest.java (hunk 18 above)
Comment (igorchupin) [reply src/test/java/LoginLogoutTest.java]: This was changed

File: src/test/java/LoginLogoutTest.java (hunk 19 above)
Comment (igorchupin) [reply src/test/java/LoginLogoutTest.java]: This was removed

File: src/test/java/LoginLogoutTest.java (hunk 20 above)
Comment (igorchupin) [reply src/test/java/LoginLogoutTest.java]: This was changed

File: src/test/java/LoginLogoutTest.java (hunk 21 above)
Comment (igorchupin) [reply src/test/java/LoginLogoutTest.java]: This was changed

File: src/test/java/LoginLogoutTest.java (hunk 22 above)
Comment (igorchupin) [reply src/test/java/LoginLogoutTest.java]: This was changed

File: src/main/java/ChooseAccountPage.java [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+
+    private final WebDriver driver;
+    private static final By YOU_ARE_LOGGED_OUT_TEXT = By.xpath("//h3[@class='AuthAccountList-unathorizedAccountsTitle']");
+    private static final By CURRENT_ACCOUNT_NAME = By.xpath("//span[@class='CurrentAccount-displayName']");
+
+    public ChooseAccountPage() {
+       driver = SingleDriver.getSingleDriverInstance().getDriver();
+    }
+
+    public String getTextThatUserWasLoggedOut() {
```
Comment (EugenBorisik) [inline src/main/java/ChooseAccountPage.java:19]: Still don't like this method name, but let it be as now

File: src/main/java/HomePage.java [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+public class HomePage {
+    private final WebDriver driver;
+    private static final By USER_ACCOUNT_NAME = By.xpath("//div[@class='legouser__menu-header']//span[@class='user-account__name']");
+    private static final By USER_ICON = By.className("user-pic__image");
+    private static final By SIGN_OUT_BUTTON = By.linkText("Log out");
+
+   public HomePage() {
+       driver= SingleDriver.getSingleDriverInstance().getDriver();
+       WebDriverWait wait = new WebDriverWait(driver, 5);
+       wait.until(ExpectedConditions.titleContains("Inbox — Yandex.Mail"));
```
Comment (EugenBorisik) [inline src/main/java/HomePage.java]: Let's replace this waiter with waiter for user icon f.e.

File: src/main/java/SingleDriver.java [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+    private WebDriver driver;
+
+    private SingleDriver() {
+         WebDriverManager.chromedriver().setup();
+         driver = new ChromeDriver();
+         driver.manage().timeouts().implicitlyWait(5, TimeUnit.SECONDS);
+    }
+
+    public static SingleDriver getSingleDriverInstance () {
+        if (singleDriver == null) {
```
Comment (EugenBorisik) [inline src/main/java/SingleDriver.java:18]: In previous version you had double-check locking, please return it.
Now your singleton looks like singleton :)

File: src/test/java/LoginLogoutTest.java [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+    private static final String EXPECTED_NAME = "Inbox — Yandex.Mail";
+
+    @BeforeEach
+    public void beforeTest () {
+        driver = SingleDriver.getSingleDriverInstance().getDriver();
+    }
+
+    @AfterEach
+    public void afterTest()  {
+     SingleDriver.getSingleDriverInstance().closeDriver();
```
Comment (EugenBorisik) [inline src/test/java/LoginLogoutTest.java:20]: Format code

File: src/test/java/LoginLogoutTest.java [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+    @Test
+    @Tag("yandex_logout_test")
+    @DisplayName("Logout from mail test")
+    void yandexLogoutTest() {
+       HomePage homePage = loginPage.loginToMail(username, password);
+       ChooseAccountPage chooseAccountPage = homePage.signOut();
+
+        assertAll("Choose account Page",
+                () -> assertTrue(chooseAccountPage.getTextThatUserWasLoggedOut().matches("You are logged out of:"),
+                        "Incorrect page was opened (Incorrect Title)"),
```
Comment (EugenBorisik) [inline src/test/java/LoginLogoutTest.java]: Correct this message, also in this case make sense to add actual text in message

File: src/test/java/LoginLogoutTest.java [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Test
+    @Tag("yandex_login_test")
+    @DisplayName("Login in mail test")
+    void yandexLoginTest() {
+        HomePage homePage = loginPage.loginToMail(username, password);
+
+        assertAll("HomePage",
+                () -> assertTrue(driver.getTitle().contains(EXPECTED_NAME),
+                        "Incorrect page was opened (Incorrect Title)"),
```
Comment (EugenBorisik) [inline src/test/java/LoginLogoutTest.java]: It makes sense to add actual title in message

**Review by EugenBorisik — CHANGES_REQUESTED** [review]

File: src/main/java/ChooseAccountPage.java (hunk 23 above)
Comment (igorchupin) [reply src/main/java/ChooseAccountPage.java:19]: =((( I have no other idea

File: src/main/java/HomePage.java (hunk 24 above)
Comment (igorchupin) [reply src/main/java/HomePage.java]: This was changed
commit b42cf9e76fc77e1fd184e8564df1c4bca8e33144

File: src/main/java/SingleDriver.java (hunk 25 above)
Comment (igorchupin) [reply src/main/java/SingleDriver.java:18]: This was returned
commit b42cf9e76fc77e1fd184e8564df1c4bca8e33144

File: src/test/java/LoginLogoutTest.java (hunk 26 above)
Comment (igorchupin) [reply src/test/java/LoginLogoutTest.java:20]: Done
commit b42cf9e76fc77e1fd184e8564df1c4bca8e33144

File: src/test/java/LoginLogoutTest.java (hunk 28 above)
Comment (igorchupin) [reply src/test/java/LoginLogoutTest.java]: Now it shows the current title
commit b42cf9e76fc77e1fd184e8564df1c4bca8e33144

File: src/test/java/LoginLogoutTest.java (hunk 27 above)
Comment (igorchupin) [reply src/test/java/LoginLogoutTest.java]: The message was changed to more user-friendly.
commit b42cf9e76fc77e1fd184e8564df1c4bca8e33144

**Review by EugenBorisik — APPROVED** [review]

**Pull request merged**